### PR TITLE
Notification flow control

### PR DESCRIPTION
**Pull request by fgalan**
Implements #3568

Doc pending:

- [x] CLI doc
- [x] NGSIv2 notes doc
- [x] Perf tunning doc
- [ ] ~Devel manual?~ at the end, I think this is not necessary as the additions to mongoUpdateContext() are somehow a "side functionality", outside of the main flow described in the devel document

### CODE REVIEW

File: src/lib/mongoBackend/mongoUpdateContext.cpp [hunk 1]
```diff
@@ -38,12 +38,60 @@
...
+  // get current size of the queue
+  unsigned int qi = ((QueueNotifier*) getNotifier())->queueSize();
+#endif
+  LM_T(LmtNotifier, ("flow control pass %d, delay %d, notification queue size is: %d",
+                     pass, accumulatedDelay, qi));
+
+
+  while (qi > target)
+  {
+    pass++;
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoUpdateContext.cpp]: We should check that the connection with the client is still active or break in negative case. Not sure how to do this... we have a ciP->connection object (a MDH_Connection) which hopefully could be used to check this. However libmicrohttpd library documentation is not very clear on this. Maybe the MHD_get_connection_info() function can be used for this.

Comment (fgalan) [reply src/lib/mongoBackend/mongoUpdateContext.cpp]: It is not clear how to implement this at the presesnt moment (suggestions/hints welcome ;) so probably I'll move the text to a code comment with a FIXME mark. It is not a big issue, after all (only involves some inefficiency in the case of client cutting the connection, which is not usual).

Comment (fgalan) [reply src/lib/mongoBackend/mongoUpdateContext.cpp]: Comment added in commit 612b80d12a88236b1927641547d83491d95fb3ba

File: doc/manuals/admin/perf_tuning.md [hunk 2]
```diff
@@ -122,6 +123,60 @@ on [the `notifQueue` block](statistics.md#notifqueue-block) may help you to tune
...
+
+Orion implements a flow control mechanism that allow to slow-down the updates flow sent by a client.
+Using flow control, Orion does not respond immediately to update request, but when the notifications
+triggered by the update has been sent (totally or partially, depending on the flow control
+configuration). This way, we avoid Orion saturation due to too much accumulated notifications
+in the notification queue (which, at the end, causes notifications being discarded when the queue
+is full).
+
+Flow control mechanism is configured using `-notifFlowControl` which takes three parameters:
+`-notifFlowControl gauge:stepDelay:maxInterval`. It requires threadpool notification mode
```
Comment (manucarrace) [inline doc/manuals/admin/perf_tuning.md]: Three parameters. It is not clear the meaning of each parameter and its possible values.

Comment (fgalan) [reply doc/manuals/admin/perf_tuning.md]: Actually is explained below. However, at this point the reader don't know it. I have slightly modified the text in 9a8663688bda2800ec8dc92b917cc2cb4010f447 to make it clearer.

**Comment by fgalan** [comment]
@fisuda it would be great if you could modify cli.md, perf_tunnin.md and ngsiv2_implementation_nodes.md to sync the Japanese translation, please. Thanks!

**Comment by fisuda** [comment]
Noted with thanks.

**Pull request merged**